docs/proposals/runtime-client-server.md

```diff
@@ -0,0 +1,202 @@
+
+However, this enables container runtime implementations to be built and maintained outside of the core kubernetes tree.
+
+## Proposed design
+
+**Design of client/server container runtime**
+
+The main idea of client/server container runtime is to keep main control logic in kubelet while letting remote runtime only do dedicated actions:
+
+* A remote runtime manager which implements [Runtime](../../pkg/kubelet/container/runtime.go#L74) interface will be added to kubelet
+* Keep `SyncPod()` in kubelet remote runtime manager and remote runtime only needs to start or kill a pod
+* Keep volume manager in kubelet and pass volumeMap to remote runtime when starting a new pod
+* Keep image garbage collection in kubelet and remote runtime only needs to pull or remove images
+* Merge `RunInContainer` into `ExecInContainer`
+* Avoid `api.Pod` in `RemoteRuntime` interface, but remain essential data for managing pods
+
+The workflow on how to invoke remote runtime API on Pod start is:
+
+```
+Kubelet                  RemoteRuntimeManager                RemoteRuntime
+   +                              +                               +
+   |                              |                               |
+   +---------SyncPod------------->+                               |
+   |                              |                               |
+   |                              +-----------RunPod------------->+
+   |                              +<------------------------------+
+   |                              |                               |
+   |                              XXXXXXXXXXXX                    |
+   |                              |          X                    |
+   |                              |    NetworkPlugin.             |
+   |                              |       SetupPod                |
+   |                              |          X                    |
+   |                              XXXXXXXXXXXX                    |
+   |                              |                               |
+   | <-------Success--------------+                               |
+   |                              |                               |
+   +                              +                               +
+```
+
+Furthermore, Kubelet network plugin has no sense of dedicated container runtime currently, but new container runtime may not work with existing docker or rkt plugin. So the type of container runtime and the ID of pod will be added to kubelet network plugin interface.
+
+For monitoring, cAdvisor is default monitoring solution in kubelet. But cAdvisor only supports docker and it is not working with other container runtimes. So the type of container runtime will be added to cAdvisor Interface. Each new container runtime should add their own runtime plugins to cAdvisor. This could be implemented in a few steps:
```

Comment:
I don't know whether there is a long term design for container runtime monitoring, but I think cAdvisor is extensible to new runtimes. CC @dchen1107